Generic Sensor: Extract common tests in a single file

Honry opened the pull request:
- Move all the common tests in a single file, each of the sensor
    type will call a function to get the generic tests to run.
    Specific sensor tests can still be added in their respective html files.
  - Import some chromium Layout tests

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/4386)
<!-- Reviewable:end -->

Honry commented:
@tobie, not sure if locate the file ```generic-sensor-tests.js``` in ```common``` folder is appropriate, please help correct me if I am wrong.  Thanks!

Honry committed:
Generic Sensor: Extract common tests in a single file

  - Move all the common tests in a single file, each of the sensor
    type will call a function to get the generic tests to run.
    Specific sensor tests can still be added in their respective html files.
  - Import some chromium Layout tests

tobie commented:
> @tobie, not sure if locate the file generic-sensor-tests.js in common folder is appropriate, please help correct me if I am wrong. Thanks!

I'd put tests common to all sensor specs in the `generic-sensor` folder.

Honry commented:
> I'd put tests common to all sensor specs in the generic-sensor folder.

@tobie, Updated. Thanks for your comment.

Honry committed:
Move common test file to generic-sensor folder

w3c-bots commented:
# Chrome (unstable channel) #
Testing web-platform-tests at revision ce33a1b89cab39260630fcb056c08268a78cbfd2
Using browser at version 58.0.3004.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary><a href="https://w3c-test.org/submissions/4386/gyroscope/Gyroscope.https.html">/gyroscope/Gyroscope.https.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4386/magnetometer/Magnetometer_insecure_context.html">/magnetometer/Magnetometer_insecure_context.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4386/accelerometer/Accelerometer_insecure_context.html">/accelerometer/Accelerometer_insecure_context.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4386/ambient-light/AmbientLightSensor.https.html">/ambient-light/AmbientLightSensor.https.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4386/gyroscope/Gyroscope_insecure_context.html">/gyroscope/Gyroscope_insecure_context.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4386/magnetometer/Magnetometer.https.html">/magnetometer/Magnetometer.https.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4386/ambient-light/AmbientLightSensor_insecure_context.html">/ambient-light/AmbientLightSensor_insecure_context.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4386/accelerometer/Accelerometer.https.html">/accelerometer/Accelerometer.https.html</a></summary>


|                                                        Subtest                                                         | Results |
|------------------------------------------------------------------------------------------------------------------------|---------|
|                                                                                                                        | ERROR   |
| `when construct Accelerometer with includeGravity that is set to false, the reading is different from the default one` | FAIL    |

</details>

w3c-bots commented:
# Firefox (nightly channel) #
Testing web-platform-tests at revision ce33a1b89cab39260630fcb056c08268a78cbfd2
Using browser at version BuildID 20170123125947; SourceStamp 36486fdc3813ef7943ae5b07b4128866d1938a6c
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary><a href="https://w3c-test.org/submissions/4386/gyroscope/Gyroscope.https.html">/gyroscope/Gyroscope.https.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4386/magnetometer/Magnetometer_insecure_context.html">/magnetometer/Magnetometer_insecure_context.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4386/accelerometer/Accelerometer_insecure_context.html">/accelerometer/Accelerometer_insecure_context.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4386/ambient-light/AmbientLightSensor.https.html">/ambient-light/AmbientLightSensor.https.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4386/gyroscope/Gyroscope_insecure_context.html">/gyroscope/Gyroscope_insecure_context.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4386/magnetometer/Magnetometer.https.html">/magnetometer/Magnetometer.https.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4386/ambient-light/AmbientLightSensor_insecure_context.html">/ambient-light/AmbientLightSensor_insecure_context.html</a></summary>


| Subtest | Results |
|---------|---------|
|         | ERROR   |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4386/accelerometer/Accelerometer.https.html">/accelerometer/Accelerometer.https.html</a></summary>


|                                                        Subtest                                                         | Results |
|------------------------------------------------------------------------------------------------------------------------|---------|
|                                                                                                                        | ERROR   |
| `when construct Accelerometer with includeGravity that is set to false, the reading is different from the default one` | FAIL    |

</details>

Honry commented:
@tobie, I just refresh the tests to align with latest spec, could you PTAL?

tobie commented on accelerometer/Accelerometer.https.html:
```diff
@@ -0,0 +1,55 @@
+</iframe>
+<script>
+
+function verify_sensor_reading(sensor, should_be_bull) {
```
should_be_bull?

Honry replied:
Oops, typo error, fixed.

tobie commented on accelerometer/Accelerometer.https.html:
```diff
@@ -0,0 +1,55 @@
+}
+
+function verify_both_sensor_reading(sensor1, sensor2) {
+  return sensor1.x == sensor2.x ||
```
the `||` here makes it hard to grasp what you're testing.

Honry replied:
Agree, use array and `for` loop instead.

tobie commented on accelerometer/Accelerometer.https.html:
```diff
@@ -0,0 +1,55 @@
+}
+
+async_test(t => {
+  let sensor1 = new Accelerometer({includeGravity: false});
```
Still think `includeGravity` option doesn't belong here. :)

Honry replied:
How about testing that sensor can be successfully constructed with `includeGravity` option?

tobie replied:
Yeah, that's a spec issue. The spec is mixing high- and low-level sensors up.

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+
+function runSensorFrequency(sensorType) {
+  test(() => {
+    assert_throws(new RangeError(), () => new sensorType({frequency: -60}));
```
I don't think that's in the spec. Maybe it should? Care to file against the generic sensor spec?

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+      number++;
+    };
+    sensor.onerror = t.step_func_done(event => {
+      assert_unreached(event.error.name + ":" + event.error.message);
```
you're using this construct very often. What about extracting it into it's own helper function? E.g.: 

```js
let unreached = event => {
    assert_unreached(event.error.name + ":" + event.error.message);
};

sensor.onerror = t.step_func_done(unreached);
```

Like that if you want to display something different instead, you can do it at once for all cases. Just a thought.

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+      sensor.stop();
+      t.done();
+    }, 1000);
+  }, "frequency is capped to 60.0 Hz");
```
This is clearly **not** a spec requirement and **should not be tested**.

tobie commented on gyroscope/Gyroscope.https.html:
```diff
@@ -0,0 +1,30 @@
+<script>
+
+function verify_sensor_reading(sensor, should_be_bull) {
+  if (should_be_bull) {
```
should be bull?

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+function runGenericSensorInsecureContext(sensorType) {
+  test(() => {
+    assert_throws('SecurityError', () => {
+      let sensor = new sensorType();
```
assignment is not necessary.

tobie commented on gyroscope/Gyroscope_onerror-manual.https.html:
```diff
@@ -0,0 +1,20 @@
+<h2>Precondition</h2>
+<ol>
+  <li>
+    Disable the Gyroscope Sensor or run test on a device without Gyroscope Sensor.
```
This requires a manual step and access to the device under test (to disconnect the sensor). It will fail otherwise on browsers that correctly implement the spec. Wondering if there isn't a way to indicate that. Or maybe just mark it as a manual test?

Summoning @jgraham, @Ms2ger and @gsnedders for advice here.

gsnedders replied:
Just mark it as manual.

tobie replied:
It's been a long time. Forgot those were done through the pathname.

tobie commented on magnetometer/Magnetometer_onerror-manual.https.html:
```diff
@@ -0,0 +1,20 @@
+</ol>
+<script>
+
+runGenericSensorOnerror(Magnetometer);
```
Same concern wrt to requiring a manual step.

tobie replied:
nm

tobie commented on accelerometer/Accelerometer_onerror-manual.https.html:
```diff
@@ -0,0 +1,20 @@
+</ol>
+<script>
+
+runGenericSensorOnerror(Accelerometer);
```
Same concern wrt to requiring a manual step.

Honry replied:
@gsnedders had answered the concern. :)

tobie replied:
OK, that's mentioned in path. Missed that.

tobie commented on ambient-light/AmbientLightSensor_onerror-manual.https.html:
```diff
@@ -5,28 +5,16 @@
 
-  sensor.start();
-}, "Test that 'onerror' event is fired when sensor is not supported");
+runGenericSensorOnerror(AmbientLightSensor);
```
Same concern wrt to requiring a manual step.

tobie replied:
nm

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+  }, "throw a 'SecurityError' when construct sensor in an insecure context");
+}
+
+function runGenericSensorOnerror(sensorType) {
```
Same concern wrt to requiring a manual step.

tobie commented on magnetometer/Magnetometer.https.html:
```diff
@@ -0,0 +1,30 @@
+</iframe>
+<script>
+
+function verify_sensor_reading(sensor, should_be_bull) {
```
You might want to make verify_sensor_reading more generic.

E.g. take a sensor object and an array or prop names to loop through.

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+      //After first sensor stops its reading values are null,
+      //reading values for the second sensor remains
+      sensor1.stop();
+      assert_true(verifyReading(sensor1, true /*should be null*/));
```
Can you have two methods here instead?

e.g. assert_reading_not_null(sensor1) and  assert_reading_null(sensor1).

It's super hard to understand the test otherwise.

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+      assert_unreached(event.error.name + ":" + event.error.message);
+    });
+    sensor.start();
+    t.step_timeout(() => {
```
Not sure why the timeout is necessary, here.

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+      assert_unreached(event.error.name + ":" + event.error.message);
+    };
+    //The default sensor.state is 'unconnected'
+    assert_equals(sensor.state, "unconnected");
```
So these state names are (again) about to change.

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+    //the sensor.start() return undefined
+    assert_equals(start_return, undefined);
+    sensor.stop();
+  }, "sensor.start() is correct");
```
That's vague. You might want to add more specific tests instead. e.g. "`sensor.start` doesn't throw in [STATE] state," "`sensor.start` returns undefined," "`sensor.start` triggers activation," etc.

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+    assert_equals(sensor.state, "idle");
+    //the sensor.stop() returns undefined
+    assert_equals(stop_return, undefined);
+  }, "sensor.stop() is correct");
```
Same comments as for sensor.start()

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+      assert_true(verifyReading(sensor));
+      let cachedSensor = sensor;
+      let win = window.open('', '_blank');
+      assert_equals(sensor, cachedSensor);
```
you just defined cachedSensor as sensor, what's the point of this assertion?

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -0,0 +1,158 @@
+      assert_unreached(event.error.name + ":" + event.error.message);
+    });
+    sensor.start();
+  }, "sensor readings can not be fired on the background tab");
```
I don't understand how this test functions at all.

tobie commented on magnetometer/Magnetometer_insecure_context.html:
```diff
@@ -0,0 +1,20 @@
+<h2>Precondition</h2>
+<ol>
+  <li>
+    Run test in an insecure context, e.g. http://example.com/.
```
Afaik, you can automate running the test in http vs. https.

@jgraham?

tobie replied:
nm

tobie commented on ambient-light/AmbientLightSensor_insecure_context.html:
```diff
@@ -5,19 +5,16 @@
-    let sensor = new AmbientLightSensor();
-  });
-}, "throw a 'SecurityError' when construct AmbientLightSensor in an insecure context");
+runGenericSensorInsecureContext(AmbientLightSensor);
```
Same concern. I think this can be automated. It should be.

Honry replied:
Actually it is an auto test and ran in http, I just note here to tell testers using insecure context.

tobie replied:
Apologies. I didn't know that was achieved by doing `xxx.https.html`.

Honry committed:
Adapt sensor tests to latest spec

Honry commented:
@tobie, thank you so much for the exhaustive comments, just address them, PTAL.

Honry committed:
Address @Tobie's comments

  - Extract common function into helper function
  - Remove test which is not mentioned in spec
  - Make start/stop tests more clear
  - Fix typo and scrap issues

tobie commented on generic-sensor/generic-sensor-tests.js:
```diff
@@ -138,21 +204,4 @@ function runSensorFrequency(sensorType) {
   test(() => {
     assert_throws(new RangeError(), () => new sensorType({frequency: -60}));
```
Consensus seems to be not to throw: https://github.com/w3c/sensors/issues/169

Honry replied:
Yes, I'd like to remove this test.

Honry committed:
Remove negative frequency test

w3c-bots commented:
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/4386/)

Honry commented:
@tobie, I just refresh the tests to make them compliance with latest spec, could you please help check again?

tobie reviewed (approved):
LGTM

Honry commented:
@tobie, thank you so much for the time on reviewing this PR.

Honry committed:
Adapt tests to latest spec

 - Stop exposing Sensor state
 - Drop includesGravity

Honry committed:
Remove nested assertion
